swaybar/tray/item.c

```diff
@@ -493,24 +493,36 @@ uint32_t render_sni(cairo_t *cairo, struct swaybar_output *output, double *x,
 		cairo_destroy(cairo_icon);
 	}
 
-	int padded_size = icon_size + 2*padding;
-	*x -= padded_size;
-	int y = floor((height - padded_size) / 2.0);
+	double descaled_padding = (double)padding / output->output_scale;
+	double descaled_icon_size = (double)icon_size / output->output_scale;
+
+	int size = descaled_icon_size + 2 * descaled_padding;
+	*x -= size;
+	int icon_y = floor((output->height - size) / 2.0);
 
 	cairo_operator_t op = cairo_get_operator(cairo);
 	cairo_set_operator(cairo, CAIRO_OPERATOR_OVER);
-	cairo_set_source_surface(cairo, icon, *x + padding, y + padding);
```

Comment:
`cairo_set_source_surface` doesn't allow control of the CTM scaling directly, but the underlying pattern it obscures does allow it and could be retrieved and manipulated.  Instead, just do what `cairo_set_source_surface` does ourselves and setup the CTM as we desire (now with scaling).

---

https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-set-source-surface
>This is a convenience function for creating a pattern from surface and setting it as the source in cr with cairo_set_source().

https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-set-source
>Note: The pattern's transformation matrix will be locked to the user space in effect at the time of cairo_set_source(). **This means that further modifications of the current transformation matrix will not affect the source pattern.** See cairo_pattern_set_matrix().